### PR TITLE
Support multiple hubs

### PR DESCRIPTION
**Pull request by msenmurugan**

### CODE REVIEW

**Review by sig-fisherj — APPROVED** [review]
👍 💯

File: pkg/core/hubmanager.go [hunk 1]
```diff
@@ -96,10 +97,10 @@ func NewHubManager(newHub hubClientCreator, stop <-chan struct{}) *HubManager {
 }
 
 // SetHubs ...
-func (hm *HubManager) SetHubs(hubURLs []string) {
+func (hm *HubManager) SetHubs(hubs map[string]*Host) {
 	newHubURLs := map[string]bool{}
-	for _, hubURL := range hubURLs {
-		newHubURLs[hubURL] = true
+	for _, hub := range hubs {
```
Comment (rrati) [inline pkg/core/hubmanager.go]: You don't need to do this anymore.  The original code was converting an array into a map for a simple test for existence.  SetHubs now is receiving a map, so should be able to use the map that is passed in the same way.

File: pkg/core/hubmanager.go [hunk 2]
```diff
@@ -110,10 +111,13 @@ func (hm *HubManager) SetHubs(hubURLs []string) {
 	// 1. create new hubs
 	// TODO handle retries and failures intelligently
 	go func() {
-		for hubURL := range hubsToCreate {
-			err := hm.create(hubURL)
-			if err != nil {
-				log.Errorf("unable to create Hub client for %s: %s", hubURL, err.Error())
+		for host := range hubsToCreate {
+			if _, ok := hm.hubs[host]; !ok {
```
Comment (rrati) [inline pkg/core/hubmanager.go]: Can combine this and next line as:
 
if hub, ok := hm.hubs[host]; ok {

File: pkg/core/perceptor.go [hunk 3]
```diff
@@ -167,6 +175,22 @@ func NewPerceptor(config *Config, timings *Timings, scanScheduler *ScanScheduler
 	return perceptor, nil
 }
 
+// getBlackDuckHosts will get the list of Black Duck hosts
+func getBlackDuckHosts(config *Config) (map[string]*Host, error) {
+	password, ok := os.LookupEnv(config.BlackDuck.ConnectionsEnvironmentVariableName)
```
Comment (rrati) [inline pkg/core/perceptor.go]: Same as the model version of the function.  Just confusing variable?

**Review by rrati — COMMENTED** [review]
If you could clean up the comments for the structs/functions you touched where there is the ... as the description, that would be great.

File: pkg/core/hubmanager.go (hunk 1 above)
Comment (msenmurugan) [reply pkg/core/hubmanager.go]: Fixed it

File: pkg/core/config.go [hunk 4]
```diff
@@ -82,20 +90,42 @@ type PerceptorConfig struct {
...
-func (config *Config) model() *api.ModelConfig {
+// getModelBlackDuckHosts will get the list of Black Duck hosts
+func (config *Config) getModelBlackDuckHosts() ([]*api.ModelHost, error) {
+	password, ok := os.LookupEnv(config.BlackDuck.ConnectionsEnvironmentVariableName)
+	if !ok {
+		return nil, fmt.Errorf("cannot find Black Duck hosts: environment variable %s not found", config.BlackDuck.ConnectionsEnvironmentVariableName)
+	}
+
+	blackduckHosts := map[string]*api.ModelHost{}
+	err := json.Unmarshal([]byte(password), &blackduckHosts)
```
[the remark this answers is not in the document]
Comment (msenmurugan) [reply pkg/core/config.go]: renamed password to connectionStrings

File: pkg/core/hubmanager.go (hunk 2 above)
Comment (msenmurugan) [reply pkg/core/hubmanager.go]: We can't do this. Bcoz both are from different struct

File: pkg/core/perceptor.go (hunk 3 above)
Comment (msenmurugan) [reply pkg/core/perceptor.go]: renamed password to connectionstrings

File: api/swagger-spec.json [hunk 5]
```diff
@@ -411,16 +416,37 @@
           "description": "The SHA of the image",
           "type": "string"
         },
-        "HubProjectName": {
-          "description": "The Hub project name",
+        "Scheme": {
+          "description": "The Scheme of the Black Duck instance. Currently, supported only https",
```
[the remark this answers is not in the document]
Comment (msenmurugan) [reply api/swagger-spec.json]: done

File: api/swagger-spec.json [hunk 6]
```diff
@@ -411,16 +416,37 @@
...
           "type": "string"
         },
-        "HubProjectVersionName": {
-          "description": "The Hub project version name",
+        "Domain": {
+          "description": "The Domain or Host name or IP address of the Black Duck instance",
```
[the remark this answers is not in the document]
Comment (msenmurugan) [reply api/swagger-spec.json]: done

File: api/swagger-spec.json [hunk 7]
```diff
@@ -411,16 +416,37 @@
...
         },
-        "HubScanName": {
-          "description": "The Hub scan name",
+        "Port": {
+          "description": "The Port of the Black Duck instance",
+          "type": "integer",
+          "format": "int"
+        },
+        "User": {
+          "description": "The User name of the Black Duck instance",
```
[the remark this answers is not in the document]
Comment (msenmurugan) [reply api/swagger-spec.json]: done

File: api/swagger-spec.json [hunk 8]
```diff
@@ -411,16 +416,37 @@
...
+          "type": "string"
+        },
+        "Password": {
+          "description": "The Password of the Black Duck instance",
```
[the remark this answers is not in the document]
Comment (msenmurugan) [reply api/swagger-spec.json]: done

File: api/swagger-spec.json [hunk 9]
```diff
@@ -411,16 +416,37 @@
...
+          "type": "string"
+        },
+        "BlackDuckProjectName": {
+          "description": "The Black Duck project name",
```
[the remark this answers is not in the document]
Comment (msenmurugan) [reply api/swagger-spec.json]: done

File: api/swagger-spec.json [hunk 10]
```diff
@@ -580,8 +606,8 @@
           "description": "The number of scans that may be run at the same time",
           "type": "integer"
         },
-        "HubClientTimeoutMilliseconds": {
-          "description": "The timeout for HTTP requests issued to the Hub",
+        "ClientTimeoutMilliseconds": {
+          "description": "The timeout for HTTP requests issued to the Black Duck",
```
[the remark this answers is not in the document]
Comment (msenmurugan) [reply api/swagger-spec.json]: done

**Review by rrati — APPROVED** [review]

**Pull request merged**